File: lib/core/utils/stable_hash.dart
import 'dart:convert';

/// FNV-1a 32-bit hash — deterministic across Dart/JS/Wasm and reproducible
/// server-side, unlike `String.hashCode` (which the Dart language spec does
/// NOT guarantee stable across runtimes/restarts).
///
/// Replaces `'$feature:$uid'.hashCode` in `AppConfigService.isInRollout`
/// (see DECISIONS.md ADR-023). The real defect that motivated this wasn't
/// instability across app restarts — it was that a Cloud Function could
/// never reproduce the same rollout bucket a client computed, so a user
/// could end up inside a gradual rollout on the client and outside it on
/// the server for the same feature. Mirrored exactly in
/// `functions/stable_hash.js`; both are asserted against the SAME vector
/// table in their respective test files — that shared table is what keeps
/// the two implementations honest.
int fnv1a32(String input) {
  const fnvOffsetBasis = 0x811c9dc5;
  const fnvPrime = 0x01000193;
  var hash = fnvOffsetBasis;
  for (final byte in utf8.encode(input)) {
    hash ^= byte;
    hash = (hash * fnvPrime) & 0xFFFFFFFF;
  }
  return hash;
}

File: functions/stable_hash.js
'use strict';

// FNV-1a 32-bit hash — mirrors lib/core/utils/stable_hash.dart exactly.
// See that file's doc comment (and PLAN.md Faz A §A3) for why this exists:
// a Cloud Function must be able to reproduce the exact same rollout bucket
// a client computed for `AppConfigService.isInRollout`, which Dart's
// String.hashCode cannot guarantee.

const FNV_OFFSET_BASIS = 0x811c9dc5;
const FNV_PRIME = 0x01000193;

function fnv1a32(input) {
  let hash = FNV_OFFSET_BASIS;
  const bytes = Buffer.from(input, 'utf8');
  for (let i = 0; i < bytes.length; i++) {
    hash ^= bytes[i];
    hash = Math.imul(hash, FNV_PRIME) >>> 0;
  }
  return hash;
}

module.exports = { fnv1a32 };
